src/hotspot/share/runtime/thread.hpp

```diff
@@ -1328,7 +1322,7 @@ class JavaThread: public Thread {
     // we have checked is_external_suspend(), we will recheck its value
     // under SR_lock in java_suspend_self().
     return (_special_runtime_exit_condition != _no_async_condition) ||
-            is_external_suspend() || is_trace_suspend() || is_obj_deopt_suspend();
+           (_suspend_flags & (_external_suspend | _obj_deopt JFR_ONLY(| _trace_flag))) != 0;
```

Comment:
I don't see this change mentioned in the bug report and it's
not obvious why you're changing this code. Also, I think this
is going to conflict with @robehn fix for JDK-8257831.

Reply:
Unless there is some significant performance issue with the lack of inlining I'd prefer to maintain code readability. And with Robbin's change is the inlining simplified anyway? Otherwise perhaps we should (re-)introduce is_any_suspend() to encapsulate the _suspend_flag checking?

Reply:
IMHO it is straight forward to check for multiple bits using a mask.
I like that this improves fast paths e.g. of JNI calls.

Reply:
With Robbin's change is_external_suspend() will be gone since that check would be already embedded in the poll check. But because _suspend_flags is declared volatile the compilers will still generate two different checks for is_trace_suspend() || is_obj_deopt_suspend() instead of doing just one (checked that with gcc 10.2 and clang 12; without volatile only one check is made). After Robbin's change that line would become (_suspend_flags & (_obj_deopt JFR_ONLY(| _trace_flag))) != 0. I also would prefer this instead of having to jump to a different method but I don't mind encapsulating them.

Reply:
I can live with it, but I highly doubt this has any observable affect on performance - especially taking into consideration the other code path changes that have been introduced in this change.